tinygrad/tensor.py

```diff
@@ -835,11 +835,14 @@ def pow(self, x:Union[Tensor, Scalar], reverse=False) -> Tensor:
     return ar.mul(sign * base_sign + (1 - base_sign)).mul(inject_nan)
   def xor(self, x:Tensor, reverse=False) -> Tensor: return mlops.Xor.apply(*self._broadcasted(x, reverse))
 
-  # TODO: this implicitly changes dtype with /2
-  def maximum(self, x:Union[Tensor, Scalar]) -> Tensor: return (self<x).detach().where(x, (self>x).detach().where(self, (self+x)/2))
-  def minimum(self, x:Union[Tensor, Scalar]) -> Tensor: return -((-self).maximum(-x))
+  def maximum(self, x:Union[Tensor, Scalar]) -> Tensor:
+    return (self<x).detach().where(x, (self>x).detach().where(self, (self+x)/2) if self.requires_grad and self.is_floating_point() else self)
```

Comment:
use both `requires_grad` and `is_floating_point` do not make sense because `bool` and `int` should not have `requires_grad` and we should check that somewhere

i don't like branching based on implementation details though and it's fine to just cast at the end. with 2 wheres it's already pretty complicated

let's also keep minimum to reuse maximum, unless the removed neg makes it significantly faster. code reuse is less likely to introduce new bug in the future and easier to maintain. the neg should be constant fused so i would be surprised if the speed is very different

Reply:
>  unless the removed neg makes it significantly faster.

it's barely faster. Unnoticable.
I tested it before and I was wondering why :D
You're right. Maintainability! 

> i don't like branching based on implementation details though and it's fine to just cast at the end. with 2 wheres it's already pretty complicated

makes sense. My initial idea was that the second `.where` and the `div` is only there just to apply the backward pass, and we only backward pass when `self` is `requires_grad`. (`self` has to be `div(2)` where `self == x` for backward)